lib/tasks/migrate/20230224205948_add_legacy_smooch_data_to_tipline_message.rake

```diff
@@ -0,0 +1,213 @@
+
+      remaining_team_ids.each_with_index do |team_id, index|
+        puts "[#{Time.now}] Creating historic tipline messages for team with ID #{team_id}. (#{index + 1} / #{remaining_team_ids.length})"
+
+        team = Team.find(team_id)
+
+        # Messages sent to tipline from users, and estimated responses from tiplines
+        MigratedTiplineMessageHelper.project_media_requests(team_id, cutoff_time).in_batches do |batch|
+          tipline_messages = batch.map { |req| MigratedTiplineMessageHelper.approximate_conversations(team_id, req) }.flatten
+          result = MigratedTiplineMessage.import(tipline_messages, on_duplicate_key_ignore: true)
+          puts "[#{Time.now}] Saved #{result.ids.count} of #{tipline_messages.size} in this batch of project media requests"
+        end
+        MigratedTiplineMessageHelper.team_requests(team_id, cutoff_time).in_batches do |batch|
+          tipline_messages = batch.map { |req| MigratedTiplineMessageHelper.approximate_conversations(team_id, req) }.flatten
+          result = MigratedTiplineMessage.import(tipline_messages, on_duplicate_key_ignore: true)
+          puts "[#{Time.now}] Saved #{result.ids.count} of #{tipline_messages.size} in this batch of team requests"
+        end
+
+        # Reports sent to users
+        MigratedTiplineMessageHelper.reports_received(team_id, cutoff_time).in_batches do |batch|
+          reports = batch.map do |f|
+            annotation = f.is_a?(Annotation) ? f : f.annotation
+            smooch_data = begin JSON.parse(annotation.load.get_field_value('smooch_data')) rescue next end
+            MigratedTiplineMessage.from_smooch_annotation_data(smooch_data, external_id: "#{smooch_data["_id"]}_report", direction: :outgoing, event: 'fact_check_report_annotation', team_id: team_id)
+          end.compact
+          result = MigratedTiplineMessage.import(reports, on_duplicate_key_ignore: true)
+          puts "[#{Time.now}] Saved #{result.ids.count} of #{reports.size} in this batch of reports sent"
+        end
+
+        # Newsletters sent to users
+        tbi = TeamBotInstallation.where(team: team, user: BotUser.smooch_user).last
+        if tbi
+          sent_newsletters_by_language = {}
+          # Gather all sent newsletters, grouped by language
+          Version.from_partition(team_id).
+            where(whodunnit: BotUser.smooch_user.id.to_s, item_id: tbi.id.to_s, item_type: ['TeamUser', 'TeamBotInstallation']).
+            where("created_at < ?", cutoff_time).each do |newsletter_version|
+              begin
+                YAML.load(JSON.parse(newsletter_version.object_after)['settings'])['smooch_workflows'].each do |smooch_workflow|
+                  language = smooch_workflow['smooch_workflow_language']
+                  sent_time = smooch_workflow['smooch_newsletter']['smooch_newsletter_last_sent_at']
+
+                  sent_newsletters_by_language[language] ||= []
+                  sent_newsletters_by_language[language] << sent_time.to_s
```

Comment:
Not every version will have a `smooch_newsletter_last_sent_at`, so let's add only if it's a valid timestamp.